### PR TITLE
frontend: tackling server state complexity in frontend development

### PR DESCRIPTION
**Pull request by zlatanpham**

### CODE REVIEW

File: Engineering/Frontend/Thought on global state management.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+---
+tags: engineering/frontend, frontend, state-
```
Comment (trankhacvy) [inline Engineering/Frontend/Thought on global state management.md:2]:
```suggestion
tags: engineering/frontend, frontend, state-management
```

**Comment by trankhacvy** [comment]
In my opinion, using swr or react-query as examples of state management is not really reasonable. It's just our way to handle the state in our app. These libraries were not created for the purpose of state management. 
I think we should try more "real" state management solutions and bring them to this discussion.

**Comment by zlatanpham** [comment]
> In my opinion, using swr or react-query as examples of state management is not really reasonable. It's just our way to handle the state in our app. These libraries were not created for the purpose of state management. I think we should try more "real" state management solutions and bring them to this discussion.

The brainery doesn't aim to discuss a state management approach but its focus is on a common wrong view of global state management. So I think it would be fine.

File: Engineering/Frontend/Thought on global state management.md (hunk 1 above)
Comment (zlatanpham) [reply Engineering/Frontend/Thought on global state management.md]: Fixed

**Comment by trankhacvy** [comment]
> The brainery doesn't aim to discuss a state management approach but its focus is on a common wrong view of global state management. So I think it would be fine.

I got your point but using the term `global state management` is a bit confusing.
For example:
- To manage user session (it's a global state) in a react app, react context (or redux) is a better choice than swr/react-query
- Using swr/react-query to fetch the data for a specific page. This data is only used inside this page so we can't consider its global state.

I think we can consider changing the title to `server state` rather than `global state`

**Comment by trankhacvy** [comment]
P/s: redux has its own tooling for data fetching and caching tool. Should we completely remove swr/react-query and use redux toolkit instead? =))

https://redux-toolkit.js.org/rtk-query/overview#motivation

**Comment by zlatanpham** [comment]
> I got your point but using the term `global state management` is a bit confusing.

I named the title "Thought on global state management" and my purpose is to point out the mistake that we used to put server state to be managed inside a global store.

**Comment by zlatanpham** [comment]
> P/s: redux has its own tooling for data fetching and caching. Should we completely remove swr/react-query and use redux toolkit instead? =))

Yes, it had and we can validate if it's good. They are libraries and as long as they help us be more productive we can consider them.

**Comment by ngolapnguyen** [comment]
Correct me if I'm wrong but the general purpose of this note is to encourage the use of `swr` or `react-query` to fetch, cache & revalidate asynchronous data _continuously_ instead of trying to fetch once & manage those data globally (with `redux`), right?

In that case I kind of agree with @trankhacvy about how the title can be a little confusing. The focus here is on "asynchronous data" (or "asynchronous state"?) so I think we can change the title to be more detailed & relevant. How about "Thought on handling asynchronous data with global state magement"? 🤔

**Comment by zlatanpham** [comment]
Ok right. I'm convinced. @trankhacvy @ngolapnguyen how about if the title is "Global state vs. server-state data: Understanding the differences and choosing the right approach"  (Asking ChatGPT)

**Comment by ngolapnguyen** [comment]
I think it's better 👍 Still can't get a good grasp of the topic just by the title (maybe because I'm not familiar with the terms), but I think that title would be more relevant to the content.

**Pull request merged**